### PR TITLE
Implement Pro version payment

### PR DESCRIPTION
**Pull request by nuttyartist**
This PR implements the payment system for the Pro version. The checkout link is in test mode, soon I'll update it to production.

### CODE REVIEW

**Review by guihkx — COMMENTED** [review]
I'm not sure how other collaborators will feel about this, but I personally have no problem with it, as long as we can still get all the features by building the app from source.

*However*, I have my doubts of how *users* (i.e. people who don't know how to build from source) will feel about this. And depending on their reaction, it's possible that you end up losing a large portion of your user base, which are mostly composed of non-tech-savvy Windows users.

But in the end, you're the owner of the project, so it's your call. :)

I have two questions, though:

Right now, all binaries built by the CI don't have access to 'PRO features', and that includes binaries for Linux users, i.e.: The AppImage, the deb packages and the rpm packages are all 'limited'.

Do you intend to keep this as-is?

On the same subject, since we maintain the Flatpak and Snap packages, do you intend to make 'PRO features' unavailable for these users as well?

---

Anyway, regarding the changes itself, I just built with `-DPRO_VERSION=ON` on my Linux machine, but 'PRO features' are still locked:

Screenshot_20230724_132057

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -78,7 +78,7 @@ jobs:
             build-type: release
 
           - os: macos-11
-            qt-version: 6.5.0
+            qt-version: 6.5.2
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply .github/workflows/macos.yml:81]: Separate commit in this branch?

File: CMakeLists.txt [hunk 2]
```diff
@@ -310,12 +313,14 @@ if(QT_VERSION VERSION_GREATER_EQUAL 6.2)
...
       src/qml/OptionItemButton.qml
+      src/qml/ProPaymentWindow.qml
       src/qml/SwitchButton.qml
       src/qml/TextButton.qml
       src/qml/ThemeChooserButton.qml
       src/qml/TodoColumnDelegate.qml
       src/qml/TodoTaskDelegate.qml
-      src/qml/ViewChooserButton.qml)
+      src/qml/ViewChooserButton.qml
+      src/qml/Utilities.js)
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply CMakeLists.txt:323]: Haha no problem, forgot about this!

File: README.md [hunk 3]
```diff
@@ -55,33 +54,46 @@ The notes database and settings file are stored in:
...
 ## Notes makes use of the following open source fonts:
 
-* Roboto
-* Source Sans Pro
-* Trykker
-* Mate
-* iA Writer Mono
-* iA Writer Duo
-* iA Writer Quattro
+- Roboto
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply README.md:93]: Right!

File: CMakeLists.txt [hunk 4]
```diff
@@ -19,6 +19,7 @@ option(GIT_REVISION
        "Append the current git revision to the app's version string" OFF)
 option(UPDATE_CHECKER
        "Enable or disable both the update checker and auto-updater" ON)
+option(PRO_VERSION "Enable or disable Notes Pro features" OFF)
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply CMakeLists.txt:22]: Oh great idea! How to add the CI invokation?

File: src/aboutwindow.cpp [hunk 5]
```diff
@@ -78,7 +56,45 @@ AboutWindow::~AboutWindow()
...
+              "Ahmed<br/><br/><strong>Designers:</strong><br/>Kevin Doyle<br/><br/>And to the "
+              "many of our beloved users who keep sending us feedback, you are an essential force "
+              "in "
+              "helping us improve, thank you!<br/><br/><strong>Notes makes use of the following "
+              "third-party libraries:</strong><br/><br/>QMarkdownTextEdit<br/>"
+#if defined(UPDATE_CHECKER)
+              "QSimpleUpdater<br/>"
+#endif
+              "QAutostart<br/>QXT<br/><br/><strong>Notes makes use of the following open source "
+              "fonts:</strong><br/><br/>Roboto<br/>Source Sans Pro<br/>Trykker<br/>Mate<br/>iA "
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply src/aboutwindow.cpp:85]: Right right!

File: docs/build_options.md [hunk 6]
```diff
@@ -10,6 +10,7 @@ You should set those while invoking CMake to build the project. See the [example
 | `USE_QT_VERSION`                           | (unset)       | `5` / `6`           | Use a specific version of Qt to build the app               |
 | `GIT_REVISION`                             | `OFF`         | `ON` / `OFF`        | Append the current git revision to the app's version string |
 | `UPDATE_CHECKER`                           | `ON`          | `ON` / `OFF`        | Enable or disable both the update checker and auto-updater  |
+| `PRO_VERSION`                              | `OFF`         | `ON` / `OFF`        | Enable or disable Notes Pro features  |
```
[the remark this answers is not in the document]
Comment (nuttyartist) [reply docs/build_options.md:13]: 👍

**Comment by nuttyartist** [comment]
@guihkx Thanks for the awesome feedback!

> However, I have my doubts of how users (i.e. people who don't know how to build from source) will feel about this. And depending on their reaction, it's possible that you end up losing a large portion of your user base, which are mostly composed of non-tech-savvy Windows users.

Do you mean that the message about building from source will confuse them? Do you think it's better to remove it?

> But in the end, you're the owner of the project, so it's your call. :)

Still apprieciate all feedback, so thanks.

> Right now, all binaries built by the CI don't have access to 'PRO features', and that includes binaries for Linux users, i.e.: The AppImage, the deb packages and the rpm packages are all 'limited'.
> 
> Do you intend to keep this as-is?
> On the same subject, since we maintain the Flatpak and Snap packages, do you intend to make 'PRO features' unavailable for these users as well?

Yes. My end goal is to make the app completely free to use (i.e. no Pro version) and only charge for what truly make sense for this kind of app - built-in sync. But untl that is implemented, considering my somewhat challenging fiscal landscape, I'm trying the best I can to moentize the Pro version to a certain extent.

> Anyway, regarding the changes itself, I just built with -DPRO_VERSION=ON on my Linux machine, but 'PRO features' are still locked:

That's weird. I'll look into that.

**Comment by nuttyartist** [comment]
I think I solved all the issuesyou pointed at except, what to do about the Qt version in the macOS build? Revert it to 6.5.0?

File: docs/build_options.md [hunk 7]
```diff
@@ -10,6 +10,7 @@ You should set those while invoking CMake to build the project. See the [example
 | `USE_QT_VERSION`                           | (unset)       | `5` / `6`           | Use a specific version of Qt to build the app               |
 | `GIT_REVISION`                             | `OFF`         | `ON` / `OFF`        | Append the current git revision to the app's version string |
 | `UPDATE_CHECKER`                           | `ON`          | `ON` / `OFF`        | Enable or disable both the update checker and auto-updater  |
+| `PRO_VERSION`                              | `OFF`         | `ON` / `OFF`        | Enable or disable Notes Pro features                        |
```
Comment (guihkx) [inline docs/build_options.md:13]: This is now `ON` by default.

File: .github/workflows/macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (guihkx) [reply .github/workflows/macos.yml:81]: Yes. Don't worry, I'll split it into a different commit before the PR gets merged.

File: CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (guihkx) [reply CMakeLists.txt]: You'll have to go through each `.yml` file in `.github/workflows`, look all places where CMake is invoked, and then add the `-DPRO_VERSION=OFF` as an argument to `cmake`.

**Review by guihkx — COMMENTED** [review]
>Do you mean that the message about building from source will confuse them? Do you think it's better to remove it?

Hahah, no that's not what I meant. I was just expressing concern that some users might get annoyed by this change and switch to a different program. So the message can stay and hopefully that doesn't happen. :crossed_fingers:  

>Still apprieciate all feedback, so thanks.

No problem. :)

>But untl that is implemented, considering my somewhat challenging fiscal landscape, I'm trying the best I can to moentize the Pro version to a certain extent.

Gotcha. So, as a reminder to yourself, when you release a new version, you'll have to add `-DPRO_VERSION=OFF` to our [Flatpak manifest](https://github.com/flathub/io.github.nuttyartist.notes/blob/5be23d9d2e9b6840d8186d899db4bd87b43ab214/io.github.nuttyartist.notes.yaml#L15-L17) as well.

>I think I solved all the issues you pointed at except, what to do about the Qt version in the macOS build? Revert it to 6.5.0?

No need to. I just think it's a change that needs its own separate commit, since it's not directly related to the feature you're adding. It's usually done by [rebasing](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) this branch against `master` and then force-pushing. But if you can't figure out how to do it, I can do it later...

It'd be good if you could learn rebasing though, just so you could keep your commits as clean as possible, and as quick as possible too, just so the list of commits doesn't get too large before we inevitably need to clean it up before the PR is finally merged.

File: .github/workflows/macos.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (guihkx) [reply .github/workflows/macos.yml:81]: Yes. If you don't figure out how to do it, I can do it later, before the PR is merged.

File: docs/build_options.md (hunk 7 above)
Comment (nuttyartist) [reply docs/build_options.md:13]: Yes, as you suggested, right?

Comment (guihkx) [reply docs/build_options.md:13]: Yes, but the default value needs to be updated in this documentation as well (it's still `OFF`). 😄

Comment (nuttyartist) [reply docs/build_options.md:13]: Ohh right hah

**Comment by nuttyartist** [comment]
> Hahah, no that's not what I meant. I was just expressing concern that some users might get annoyed by this change and switch to a different program. So the message can stay and hopefully that doesn't happen. 

Oh okay, gotcha. I hope so too.

> Gotcha. So, as a reminder to yourself, when you release a new version, you'll have to add -DPRO_VERSION=OFF to our [Flatpak manifest](https://github.com/flathub/io.github.nuttyartist.notes/blob/5be23d9d2e9b6840d8186d899db4bd87b43ab214/io.github.nuttyartist.notes.yaml#L15-L17) as well.

👍 

> No need to. I just think it's a change that needs its own separate commit, since it's not directly related to the feature you're adding. It's usually done by [rebasing](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) this branch against master and then force-pushing. But if you can't figure out how to do it, I can do it later...
> 
> It'd be good if you could learn rebasing though, just so you could keep your commits as clean as possible, and as quick as possible too, just so the list of commits doesn't get too large before we inevitably need to clean it up before the PR is finally merged.

I promise I will study git properly tafter this release. 😅

**Comment by nuttyartist** [comment]
@guihkx Should it be `-DPRO_VERSION='OFF'`or `-DPRO_VERSION=OFF`? And on Windows, rather than
```
-DPRO_VERSION='OFF' \
```
Should it be 
```
-DPRO_VERSION='OFF' `
```

**Comment by guihkx** [comment]
> @guihkx Should it be `-DPRO_VERSION='OFF'`or `-DPRO_VERSION=OFF`?

I think both variants will work the same, but I personally choose no quotes.

> And on Windows, rather than
> 
> ```
> -DPRO_VERSION='OFF' \
> ```
> 
> Should it be
> 
> ```
> -DPRO_VERSION='OFF' `
> ```

Yes.

**Comment by nuttyartist** [comment]
> image
> 
> I think the size of the font on these two links at the bottom of the window could be just a tiny bit bigger. And since they're clickable links, the mouse cursor could be changed to a pointing hand. And maybe use a different color for the text?

Done. Let me know if you think we can merge this.

**Pull request merged**